#Learning Objectives

# Exploratory Data Analysis

- Objective 01 - submit assignments and save work via GitHub
- Objective 02 - load a CSV dataset from a URL using `pandas.read_csv`
- Objective 03 - load a CSV dataset from a local file using `pandas.read_csv
- Objective 04 - use basic Pandas functions for Exploratory Data Analysis-EDA
- Objective 05 - describe and discriminate between basic data types such as categorical, quantitative, continuous, discrete, ordinal, nominal and identifier



#Explore the University of California - Irvine Adult Dataset

###1) Load a dataset via its URL.

Load the Adult dataset from: 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'

Specify the column headers as:'age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week','native-country', 'income']

In [89]:
#Code here
import pandas as pd

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'
header = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week','native-country', 'income']
df = pd.read_csv(url, names = header)
df

age          workclass  fnlwgt  ... hours-per-week  native-country  income
0       39          State-gov   77516  ...             40   United-States   <=50K
1       50   Self-emp-not-inc   83311  ...             13   United-States   <=50K
2       38            Private  215646  ...             40   United-States   <=50K
3       53            Private  234721  ...             40   United-States   <=50K
4       28            Private  338409  ...             40            Cuba   <=50K
...    ...                ...     ...  ...            ...             ...     ...
32556   27            Private  257302  ...             38   United-States   <=50K
32557   40            Private  154374  ...             40   United-States    >50K
32558   58            Private  151910  ...             40   United-States   <=50K
32559   22            Private  201490  ...             20   United-States   <=50K
32560   52       Self-emp-inc  287927  ...             40   United-States    >50K

[32561 rows x 15 columns]

###2) Print the first 5 and the last 5 rows.

In [90]:
#Code here
df.head()

age          workclass  fnlwgt  ... hours-per-week  native-country  income
0   39          State-gov   77516  ...             40   United-States   <=50K
1   50   Self-emp-not-inc   83311  ...             13   United-States   <=50K
2   38            Private  215646  ...             40   United-States   <=50K
3   53            Private  234721  ...             40   United-States   <=50K
4   28            Private  338409  ...             40            Cuba   <=50K

[5 rows x 15 columns]

In [91]:
#Code here
df.tail()

age      workclass  fnlwgt  ... hours-per-week  native-country  income
32556   27        Private  257302  ...             38   United-States   <=50K
32557   40        Private  154374  ...             40   United-States    >50K
32558   58        Private  151910  ...             40   United-States   <=50K
32559   22        Private  201490  ...             20   United-States   <=50K
32560   52   Self-emp-inc  287927  ...             40   United-States    >50K

[5 rows x 15 columns]

###3) Determine the data type of each varaible.

In [92]:
#Code here
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

###4) Determine the dimensions of the dataset.

In [93]:
#Code here
df.shape

(32561, 15)

###5) Are there any missing values in the dataset?
Check for missing values using .isnull().sum().

In [94]:
#Code here
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

###6) Use the .describe() function in order to see the summary statistics of the numeric columns.

Are there any values shown that might be a code for missing data?  We'll learn how to change the values in a DataFrame tomorrow.

In [95]:
#Code here
df.describe()

age        fnlwgt  ...  capital-loss  hours-per-week
count  32561.000000  3.256100e+04  ...  32561.000000    32561.000000
mean      38.581647  1.897784e+05  ...     87.303830       40.437456
std       13.640433  1.055500e+05  ...    402.960219       12.347429
min       17.000000  1.228500e+04  ...      0.000000        1.000000
25%       28.000000  1.178270e+05  ...      0.000000       40.000000
50%       37.000000  1.783560e+05  ...      0.000000       40.000000
75%       48.000000  2.370510e+05  ...      0.000000       45.000000
max       90.000000  1.484705e+06  ...   4356.000000       99.000000

[8 rows x 6 columns]

Answer question here.

###7) Use the .describe(exclude='number') function to calculate summary statistics for the non-numeric columns




In [86]:
#Code here
df.describe(exclude='category')

serRace          year  ...    carMake     driver
count   34884.000000  34884.000000  ...      34884      34884
unique           NaN           NaN  ...         10        683
top              NaN           NaN  ...  Chevrolet  RickyRudd
freq             NaN           NaN  ...      12663        803
mean      468.597151   1990.027950  ...        NaN        NaN
std       259.245275      8.523917  ...        NaN        NaN
min         1.000000   1975.000000  ...        NaN        NaN
25%       247.000000   1983.000000  ...        NaN        NaN
50%       478.000000   1991.000000  ...        NaN        NaN
75%       696.000000   1998.000000  ...        NaN        NaN
max       898.000000   2003.000000  ...        NaN        NaN

[11 rows x 10 columns]

In [87]:
df.describe()

serRace          year  ...         prize      carsRace
count  34884.000000  34884.000000  ...  3.488400e+04  34884.000000
mean     468.597151   1990.027950  ...  2.704011e+04     39.406146
std      259.245275      8.523917  ...  4.455498e+04      4.348333
min        1.000000   1975.000000  ...  0.000000e+00     22.000000
25%      247.000000   1983.000000  ...  3.835000e+03     37.000000
50%      478.000000   1991.000000  ...  1.201000e+04     41.000000
75%      696.000000   1998.000000  ...  3.649625e+04     43.000000
max      898.000000   2003.000000  ...  2.277975e+06     50.000000

[8 rows x 8 columns]

In [88]:
df

serRace  year  yearRace  ...  carsRace    carMake          driver
0            1  1975         1  ...        35    Matador    BobbyAllison
1            1  1975         1  ...        35    Mercury    DavidPearson
2            1  1975         1  ...        35  Chevrolet     CecilGordon
3            1  1975         1  ...        35      Dodge      DaveMarcis
4            1  1975         1  ...        35       Ford     ElmoLangley
...        ...   ...       ...  ...       ...        ...             ...
34879      898  2003        36  ...        43  Chevrolet     MikeSkinner
34880      898  2003        36  ...        43      Dodge       JeffGreen
34881      898  2003        36  ...        43  Chevrolet  MichaelWaltrip
34882      898  2003        36  ...        43  Chevrolet    JohnAndretti
34883      898  2003        36  ...        43       Ford     MattKenseth

[34884 rows x 10 columns]

In [76]:
# alternatively to the solution above I prefer to use:-
import numpy as np
df.describe(exclude=np.number)

workclass education       marital-status  ...    sex  native-country  income
count      32561     32561                32561  ...  32561           32561   32561
unique         9        16                    7  ...      2              42       2
top      Private   HS-grad   Married-civ-spouse  ...   Male   United-States   <=50K
freq       22696     10501                14976  ...  21790           29170   24720

[4 rows x 9 columns]

###8) Calculate the number of observations in each category of the relationship variable.

In [77]:
#Code here
df.relationship.value_counts()

 Husband           13193
 Not-in-family      8305
 Own-child          5068
 Unmarried          3446
 Wife               1568
 Other-relative      981
Name: relationship, dtype: int64

###9) Select and print the occupation category.

In [78]:
#Code here
df.occupation # can also use df['occupation']

0              Adm-clerical
1           Exec-managerial
2         Handlers-cleaners
3         Handlers-cleaners
4            Prof-specialty
                ...        
32556          Tech-support
32557     Machine-op-inspct
32558          Adm-clerical
32559          Adm-clerical
32560       Exec-managerial
Name: occupation, Length: 32561, dtype: object

#Load a dataset saved on your personal computer

Go to the website http://users.stat.ufl.edu/~winner/datasets.html and scroll down to **NASCAR Race Results by Driver 1975-2003**

Download and save the .csv file to your personal computer.

###10) Upload the dataset using the following commands.

In [79]:
from google.colab import files
uploaded = files.upload()

Saving nascard.csv to nascard (1).csv


In [80]:
import io # importing here so that I can get io methods recommendation from colab below
uploaded


{'nascard.csv': b"serRace,year,yearRace,finish,start,laps,prize,carsRace,carMake,driver\r\n1,1975,1,1,1,191,12035,35,Matador,BobbyAllison\r\n1,1975,1,2,2,191,8135,35,Mercury,DavidPearson\r\n1,1975,1,3,25,184,6535,35,Chevrolet,CecilGordon\r\n1,1975,1,4,27,183,5035,35,Dodge,DaveMarcis\r\n1,1975,1,5,22,178,3835,35,Ford,ElmoLangley\r\n1,1975,1,6,35,175,2885,35,Chevrolet,JamesHylton\r\n1,1975,1,7,3,172,3185,35,Dodge,RichardPetty\r\n1,1975,1,8,21,168,2485,35,Chevrolet,GaryMatthews\r\n1,1975,1,9,33,167,2335,35,Dodge,EdNegre\r\n1,1975,1,10,10,166,2185,35,Chevrolet,HershelMcGriff\r\n1,1975,1,11,17,166,1985,35,Chevrolet,RichardChildress\r\n1,1975,1,12,24,148,1835,35,Chevrolet,DonPuskarich\r\n1,1975,1,13,6,145,2235,35,Dodge,RayElder\r\n1,1975,1,14,28,142,1635,35,Chevrolet,J.D.McDuffie\r\n1,1975,1,15,18,135,1535,35,Chevrolet,LarryEsau\r\n1,1975,1,16,11,125,1435,35,Ford,BillOsborne\r\n1,1975,1,17,15,125,1385,35,Chevrolet,ChuckWahl\r\n1,1975,1,18,29,118,1335,35,Chevrolet,BillSchmitt\r\n1,1975,1,19,1

In [82]:
#Code here

df = pd.read_csv(io.BytesIO(uploaded['nascard.csv']))
df

serRace  year  yearRace  ...  carsRace    carMake          driver
0            1  1975         1  ...        35    Matador    BobbyAllison
1            1  1975         1  ...        35    Mercury    DavidPearson
2            1  1975         1  ...        35  Chevrolet     CecilGordon
3            1  1975         1  ...        35      Dodge      DaveMarcis
4            1  1975         1  ...        35       Ford     ElmoLangley
...        ...   ...       ...  ...       ...        ...             ...
34879      898  2003        36  ...        43  Chevrolet     MikeSkinner
34880      898  2003        36  ...        43      Dodge       JeffGreen
34881      898  2003        36  ...        43  Chevrolet  MichaelWaltrip
34882      898  2003        36  ...        43  Chevrolet    JohnAndretti
34883      898  2003        36  ...        43       Ford     MattKenseth

[34884 rows x 10 columns]

###11) Print the first 10 rows of the dataset

In [83]:
#Code here
df.head(10)

serRace  year  yearRace  finish  ...  prize  carsRace    carMake          driver
0        1  1975         1       1  ...  12035        35    Matador    BobbyAllison
1        1  1975         1       2  ...   8135        35    Mercury    DavidPearson
2        1  1975         1       3  ...   6535        35  Chevrolet     CecilGordon
3        1  1975         1       4  ...   5035        35      Dodge      DaveMarcis
4        1  1975         1       5  ...   3835        35       Ford     ElmoLangley
5        1  1975         1       6  ...   2885        35  Chevrolet     JamesHylton
6        1  1975         1       7  ...   3185        35      Dodge    RichardPetty
7        1  1975         1       8  ...   2485        35  Chevrolet    GaryMatthews
8        1  1975         1       9  ...   2335        35      Dodge         EdNegre
9        1  1975         1      10  ...   2185        35  Chevrolet  HershelMcGriff

[10 rows x 10 columns]

###12) Answer the following questions for the variables prize, carMake and driver:


*   Is the variable categorical or quantitative?
*   If the variable is categorical, determine if it is ordinal, nominal or an identifier variable.
*   If the vriable is quantitative, determine if it is continuous or discrite.





#Answer here.

**Answers follow below:-**

Used the data description from the data source to answer

1) Prize is quantitative while carMake and driver is categorical.

2) carMake is nominal i.e. the data does not have any natural order to it and driver is an identified variable i.e. mere identifier variable to distinguish a driver from another.

3) Prize is continous as it can be counted with precision. For example, $ 12035 and 52 cents. Although, the data only captures the dollar figures as a whole number. 


#Data Storytelling Project Milestone

Begin looking at Data Storytelling (previously called Build Week) [project info](https://github.com/LambdaSchool/DS-Unit-1-Build) and at projects created by previous Lambda students.

[Projects — Lambda DS1!](https://drive.google.com/file/d/1RJoAqDgawzSHOJPlDQRRlfz-EnIsskpL/view)